src/OpenLoco/src/Drawing/SoftwareDrawingContext.cpp

```diff
@@ -0,0 +1,1709 @@
+                        // wrap.push_back(startLine); TODO: refactor to return pointers to line starts
+                        maxWidth = std::max(maxWidth, lastWordLineWith);
+                        // Insert line ending instead of space character
+                        *wordStart = '\0';
+                        ptr = wordStart + 1;
+                    }
+                    wrapCount++;
+                }
+            }
+
+            // Note that this is always the font used in the last line.
+            // TODO: refactor to pair up with each line, and to not use a global.
+            _currentFontSpriteBase = font;
+            return std::make_pair(maxWidth, std::max(static_cast<uint16_t>(wrapCount) - 1, 0));
+        }
+
+        // 0x004474BA
+        static void drawRectImpl(Gfx::RenderTarget& rt, int16_t left, int16_t top, int16_t right, int16_t bottom, uint32_t colour)
+        {
+            registers regs;
+            regs.ax = left;
+            regs.bx = right;
+            regs.cx = top;
+            regs.dx = bottom;
+            regs.ebp = colour;
+            regs.edi = X86Pointer(&rt);
+            call(0x004474BA, regs);
+        }
+
+        static void fillRect(Gfx::RenderTarget& rt, int16_t left, int16_t top, int16_t right, int16_t bottom, uint32_t colour)
+        {
+            drawRectImpl(rt, left, top, right, bottom, colour);
+        }
+
+        static void drawRect(Gfx::RenderTarget& rt, int16_t x, int16_t y, uint16_t dx, uint16_t dy, uint32_t colour)
+        {
+            // This makes the function signature more like a drawing application
+            drawRectImpl(rt, x, y, x + dx - 1, y + dy - 1, colour);
+        }
+
+        static void fillRectInset(Gfx::RenderTarget& rt, int16_t left, int16_t top, int16_t right, int16_t bottom, uint32_t colour, uint8_t flags)
+        {
+            registers regs;
+            regs.ax = left;
+            regs.bx = right;
+            regs.cx = top;
+            regs.dx = bottom;
+            regs.ebp = colour;
+            regs.edi = X86Pointer(&rt);
+            regs.si = flags;
+            call(0x004C58C7, regs);
+        }
+
+        static void drawRectInset(Gfx::RenderTarget& rt, int16_t x, int16_t y, uint16_t dx, uint16_t dy, uint32_t colour, uint8_t flags)
+        {
+            // This makes the function signature more like a drawing application
+            fillRectInset(rt, x, y, x + dx - 1, y + dy - 1, colour, flags);
+        }
+
+        // 0x00452DA4
+        static void drawLine(Gfx::RenderTarget& rt, int16_t left, int16_t top, int16_t right, int16_t bottom, uint32_t colour)
+        {
+            registers regs;
+            regs.ax = left;
+            regs.bx = top;
+            regs.cx = right;
+            regs.dx = bottom;
+            regs.ebp = colour;
+            regs.edi = X86Pointer(&rt);
+            call(0x00452DA4, regs);
+        }
+    } // Impl
+
+    void SoftwareDrawingContext::clear(Gfx::RenderTarget& rt, uint32_t fill)
```

Comment:
what is the reasoning behind this indirection? Why not just implement `clear` here?

Reply:
Because at the moment everything passes RenderTarget but the context does not have its own state currently, I tried to do that but ran into a road block, the plan is to do eliminate that once we can do that.